fix: Better stream and access token management

Vinzent03 opened the pull request:
## What kind of change does this PR introduce?

Bug fix

## What is the current behavior?

The behavior of the app in background depends on the platform, but by always stopping the access token refreshing, the realtime connection may keep an invalid access token and therefore the connection get closed by server. When the realtime connection gets interrupted, it gets automatically reconnected as long as it never sent some expired access token.

## What is the new behavior?

~~When a realtime channel is open, the access token keeps refreshing while the app is in background.~~

In addition, I did more management to keep realtiem and postgrest in sync for the stream method. If the postgrest request fails, I unsubscribe the realtime connection and close the stream as I see no benefit in leaving it open, as it would just be a realtime connection by then. The dev using the stream method should handle such cases. One could also argue about doing a few retries as well, though.
When the realtime connection gets a channel error, but gets resubscribed, I reload the postgrest data to ensure the data is up to date and didn't miss any updates due to the interrupted realtime connection. 

## Additional context

#1012 #705 #579

Vinzent03 commented on packages/realtime_client/lib/src/realtime_client.dart:
```diff
@@ -353,7 +358,7 @@ class RealtimeClient {
 
     for (final channel in channels) {
       if (token != null) {
-        channel.updateJoinPayload({'user_token': token});
```
The realtime-js library uses `access-token` [see here](https://github.com/supabase/realtime-js/blob/95ab09e47ed7732eb21ac31eac743704bacc29aa/src/RealtimeClient.ts#L322) instead of `user_token`. And I guess that's the correct payload. It got changed to access_token [here](https://github.com/supabase/realtime-js/commit/b3eafe820f61afdb4c663fa600c4f209757ba8b4#diff-f5a6b192eb9d6c05f61c33d8629b814ad893fbfb729bfa7af9b226e5c87d2e85)

coolusaHD commented:
Hey @Vinzent03 

This PR looks nice and maybe could also be the solution of our problem #674 .
If you can reply here or in the issue thread we would love to test the update when you finished your work ✌️ 

Thanks 🚀

dshukertjr reviewed (commented):
Thanks for this. This looks awesome. I'm a bit swamped today and maybe tomorrow, but let me take a look whenever I can.

dshukertjr commented:
@Vinzent03 BTW, have you actually ran this code on iOS and Android? I tried to fix some of the realtime issues, and found that some behaviors differ on the web sockets between those two OSs. I don't remember what exactly those differences were, but wanted to ask if you have tested on actual apps.

Vinzent03 commented:
I have no way to test this on any Apple device. I tested this on my android device and as a linux application. The stopping of refreshing the access token was especially an issue in the linux application, because the realtime connection keeps being open, when the app is in background, because there is no throttling after some time.

maxfornacon commented:
I changed the supabase_flutter package in my pubspec.yaml to the version of this PR. 
I started the macOS-Application and left it running over night with my OS in sleep mode and most likely disconnected WiFi at some point. 

I got the exception: 

```
Unhandled Exception: Exception: Realtime subscription closed
#0      SupabaseStreamBuilder._addException (package:supabase/src/supabase_stream_builder.dart:296:67)
#1      SupabaseStreamBuilder._getStreamData.<anonymous closure> (package:supabase/src/supabase_stream_builder.dart:202:9)
#2      RealtimeChannel.subscribe.<anonymous closure> (package:realtime_client/src/realtime_channel.dart:130:39)
#3      RealtimeChannel._onClose.<anonymous closure> (package:realtime_client/src/realtime_channel.dart:259:36)
#4      RealtimeChannel.trigger (package:realtime_client/src/realtime_channel.dart:694:22)
#5      RealtimeChannel.unsubscribe.onClose (package:realtime_client/src/realtime_channel.dart:575:7)
#6      RealtimeChannel.unsubscribe.<anonymous closure> (package:realtime_client/src/realtime_channel.dart:586:7)
#7      Push._matchReceive.<anonymous closure> (package:realtime_client/src/push.dart:133:17)
#8      Iterable.forEach (dart:core/iterable.dart:347:35)
#9      Push._matchReceive (package:realtime_client/src/push.dart:132:48)
#10     Push.startTimeout.<anonymous closure> (package:realtime_client/src/push.dart:96:7)
#11     RealtimeChannel.trigger (package:realtime_client/src/realtime_channel.dart:694:22)
#12     Push.trigger (package:realtime_client/src/push.dart:106:16)
#13     RealtimeChannel.unsubscribe (package:realtime_client/src/realtime_channel.dart:605:17)
#14     RealtimeClient.leaveOpenTopic (package:realtime_client/src/realtime_client.dart:372:18)
#15     RealtimeChannel.rejoin (package:realtime_client/src/realtime_channel.dart:633:12)
#16     RealtimeChannel.rejoinUntilConnected (package:realtime_client/src/realtime_channel.dart:103:7)
#17     new RealtimeChannel.<anonymous closure> (package:realtime_client/src/realtime_channel.dart:56:26)
#18     RetryTimer.scheduleTimeout.<anonymous closure> (package:realtime_client/src/retry_timer.dart:46:15)
#19     Timer._createTimer.<anonymous closure> (dart:async-patch/timer_patch.dart:18:15)
#20     _Timer._runTimers (dart:isolate-patch/timer_impl.dart:398:19)
#21     _Timer._handleMessage (dart:isolate-patch/timer_impl.dart:429:5)
#22     _RawReceivePort._handleMessage (dart:isolate-patch/isolate_patch.dart:184:12)
```

The stream did not reconnect on its own. But at least there is this exception now.

I'm gonna do some further testing soon.

maxfornacon commented:
I'd like to test the changes in my projects.

I tried to clone the repo and switched to the `fix/realtime` branch. Then I ran `melos bs` and added the package to my pubspec.yaml like this: 

```
  supabase_flutter:
    path: /Users/maxfornacon/projects/test/supabase-flutter/packages/supabase_flutter
```

But I noticed that e.g. `realtime_client` still gets installed from hosted and not using the local version. 
Can you give me a hint on how to use all packages from the local path?

coolusaHD commented:
>  Can you give me a hint on how to use all packages from the local path?

Have you tried also add the local import to your `pubspec.yaml` file
Like 

```
realtime_client:
    path: /Users/maxfornacon/projects/test/supabase-flutter/packages/realtime_client
```

maxfornacon commented:
> Have you tried also add the local import to your pubspec.yaml file

I added all necessary packages like this as `dependency_overrides`. This works I guess.

maxfornacon commented:
I started the Application on macOS, closed my MacBook and went home. When I opened it at home I saw:

```
AuthException(message: ClientException with SocketException: Failed host lookup: 'adnwhmobivhficannylf.supabase.co' (OS Error: nodename nor servname provided, or not known, errno = 8), uri=https://adnwhmobivhficannylf.supabase.co/auth/v1/token?grant_type=refresh_token, statusCode: null, errorCode: null)
flutter: #0      GotrueFetch._handleRequest (package:gotrue/src/fetch.dart:179:7)
flutter: <asynchronous suspension>
flutter: #1      GotrueFetch.request (package:gotrue/src/fetch.dart:130:12)
flutter: <asynchronous suspension>
flutter: #2      GoTrueClient._refreshAccessToken.<anonymous closure> (package:gotrue/src/gotrue_client.dart:1052:26)
flutter: <asynchronous suspension>
flutter: #3      RetryOptions.retry (package:retry/retry.dart:131:16)
flutter: <asynchronous suspension>
flutter: #4      GoTrueClient._refreshAccessToken (package:gotrue/src/gotrue_client.dart:1044:12)
flutter: <asynchronous suspension>
flutter: #5      GoTrueClient._callRefreshToken (package:gotrue/src/gotrue_client.dart:1150:20)
flutter: <asynchronous suspension>
flutter: #6      GoTrueClient._autoRefreshTokenTick (package:gotrue/src/gotrue_client.dart:1032:9)
```
This indicates a loss of internet connection, as expected.

Also the StreamBuilder printed out `RealtimeSubscribeException(status: RealtimeSubscribeStatus.channelError, details: null)`.

This is an improvement!

If it could auto reconnect would be even better. But that's something I could work with.

maxfornacon commented:
I started the Application on macOS, closed my MacBook and went home. When I opened it at home I saw:

```
AuthException(message: ClientException with SocketException: Failed host lookup: 'xxx.supabase.co' (OS Error: nodename nor servname provided, or not known, errno = 8), uri=https:/xxx.supabase.co/auth/v1/token?grant_type=refresh_token, statusCode: null, errorCode: null)
flutter: #0      GotrueFetch._handleRequest (package:gotrue/src/fetch.dart:179:7)
flutter: <asynchronous suspension>
flutter: #1      GotrueFetch.request (package:gotrue/src/fetch.dart:130:12)
flutter: <asynchronous suspension>
flutter: #2      GoTrueClient._refreshAccessToken.<anonymous closure> (package:gotrue/src/gotrue_client.dart:1052:26)
flutter: <asynchronous suspension>
flutter: #3      RetryOptions.retry (package:retry/retry.dart:131:16)
flutter: <asynchronous suspension>
flutter: #4      GoTrueClient._refreshAccessToken (package:gotrue/src/gotrue_client.dart:1044:12)
flutter: <asynchronous suspension>
flutter: #5      GoTrueClient._callRefreshToken (package:gotrue/src/gotrue_client.dart:1150:20)
flutter: <asynchronous suspension>
flutter: #6      GoTrueClient._autoRefreshTokenTick (package:gotrue/src/gotrue_client.dart:1032:9)
```
This indicates a loss of internet connection, as expected.

Also the StreamBuilder printed out `RealtimeSubscribeException(status: RealtimeSubscribeStatus.channelError, details: null)`.

This is an improvement!

If it could auto reconnect would be even better. But that's something I could work with.

maxfornacon commented:
Today I noticed the StreamBuilder not receiving new data after 5min. 

I could see this printed out from listening to the channel directly:

```
PostgresChangePayload(schema: xxx, table: notifications, commitTimestamp: 2024-09-13 12:51:23.119Z, eventType: PostgresChangeEvent.delete, newRow: {}, oldRow: {id: 55311517-01b3-4f8e-b3d5-6b4abbb43e8f}, errors: null)
```

But the StreamBuilder was not showing the change.

dshukertjr commented on packages/supabase_flutter/lib/src/supabase_auth.dart:
```diff
@@ -119,7 +119,12 @@ class SupabaseAuth with WidgetsBindingObserver {
-        Supabase.instance.client.auth.stopAutoRefresh();
+        // Realtime channels are kept alive in the background for some amount
+        // of time after the app is paused. If we stop refreshing the token
+        // here, the channels will be closed.
+        if (Supabase.instance.client.realtime.getChannels().isEmpty) {
+          Supabase.instance.client.auth.stopAutoRefresh();
+        }
```
Hmm, the purpose of this piece of code is to lower the risks of creating an invalid JWT on the client side, and we should stop attempting to refresh the token when the app is in the background. 

If we could maybe stop subscribing to realtime through stream when the app is in background, that would be a more ideal implementation of this I would say.

Vinzent03 replied:
I think it's a bad idea to disconnect from realtime when going to background. On desktop for example everything keeps running, when going to background. On mobile you may often switch for a quick moment to another app. Disconnecting and reconnecting might miss events and causes overhead. I will do some further testing on this part.

dshukertjr reviewed (commented):
The core issue with the realtime currently is the fact that the SDK isn't able to reconnect in the following scenario:
1. app is connected to realtime
2. realtime is disconnected (bad network or app goes to the background) long enough for the JWT to expire
3. app comes back online

At step 3 above, the connection should be restored, and that is the behavior of the js SDK. I have spend some time trying to fix the issue, but have had no success on this SDK, but if we could fix that, all the other issues should be fixed fairly straight forward.

Vinzent03 commented on packages/realtime_client/lib/src/constants.dart:
```diff
@@ -18,8 +18,8 @@ enum SocketStates {
   open,
 
-  /// Socket is closing.
-  closing,
```
This enum value was never in use and replaced it therefore with a new one.

Vinzent03 commented:
I've done now quite more changes. 
- I've undone #637 by again emitting the exact channel error, so that one can handle `SocketExceptions` and so on explicitly
- The errors received from `onError` aren't stringified anymore. 
- I had to make `connect` and `disconnect` async, because the previous code was wrong. We have to wait for the `conn.ready` future to be [completed before using the sink](https://pub.dev/packages/web_socket_channel#docs-and-usage).
- Added a new `SocketStates.disconnecting` since disconnecting had to be async as well, and emitting a `SocketStates.disconnected`, while it's not actually disconnected is wrong. 
- Removed `AppLifecycleState.inactive` as state to trigger stopping the refresh token and disconnecting realtime, because from reading the [documentation](https://api.flutter.dev/flutter/dart-ui/AppLifecycleState.html) this is called often and does not result in stopping the programm code
- When the app goes to actually background I now disconnect the realtime client. When resuming the work is more complicated. When quickly switching between fore and background and having a bad internet connection, the disconnection may still be running, which needed some extra care, but should be well documented in the code comments.
- I needed the method `RealtimeChannel.forceRejoin()`, because the channel is still marked as joined, which results in unsubscribing itself in `leaveOpenTopic()`, which resulted in only closed channels.

themightychris commented:
> The core issue with the realtime currently is the fact that the SDK isn't able to reconnect in the following scenario:
> 
> 1. app is connected to realtime
> 2. realtime is disconnected (bad network or app goes to the background) long enough for the JWT to expire
> 3. app comes back online
> 
> At step 3 above, the connection should be restored, and that is the behavior of the js SDK. I have spend some time trying to fix the issue, but have had no success on this SDK, but if we could fix that, all the other issues should be fixed fairly straight forward.

I consistently see the SDK fail to reconnect when the above happens well within the JWT's expiration time, is there another explanation for that addressed here?

maxfornacon commented:
Thanks @Vinzent03 for your amazing work.

I made a few tests where I had the application open for hours during the day or over night. The stream still received the changes after that, every time. This is huge!

Nevertheless, I got quite a lot of Exceptions in the log.

Mostly:
```
Unhandled Exception: RealtimeSubscribeException(status: RealtimeSubscribeStatus.channelError, details: RealtimeCloseEvent(code: 1002, reason: null))
#0      SupabaseStreamBuilder._addException (package:supabase/src/supabase_stream_builder.dart:331:67)
#1      SupabaseStreamBuilder._getStreamData.<anonymous closure> (package:supabase/src/supabase_stream_builder.dart:229:11)
#2      RealtimeChannel.subscribe.<anonymous closure> (package:realtime_client/src/realtime_channel.dart:127:39)
#3      RealtimeChannel._onError.<anonymous closure> (package:realtime_client/src/realtime_channel.dart:265:36)
#4      RealtimeChannel.trigger (package:realtime_client/src/realtime_channel.dart:718:22)
#5      RealtimeClient._triggerChanError (package:realtime_client/src/realtime_client.dart:451:15)
#6      RealtimeClient._onConnClose (package:realtime_client/src/realtime_client.dart:432:7)
#7      RealtimeClient.connect.<anonymous closure> (package:realtime_client/src/realtime_client.dart:190:11)
```

```
Unhandled Exception: RealtimeSubscribeException(status: RealtimeSubscribeStatus.channelError, details: RealtimeCloseEvent(code: 1005, reason: ))
```


What I haven't tested yet is what happens when changes occur while the application is paused. Whether they are reliably loaded when resuming.

Vinzent03 commented:
@themightychris Are you sure you've used the latest state of this pr? For me and looks like for @maxfornacon it works fine.

@maxfornacon These errors are actually expected, because the websocket connection fails when loosing internet or the app is in background. I think it's better to add them to the `.stream()`  stream, so you can react when the connection is interrupted.

themightychris commented:
@Vinzent03 I have not tested this PR in a real environment yet, I was just responding to the description of the problem that seemed to suggest JWT expiration being the root of it, which would make #674 a separate issue

maxfornacon commented:
> For me and looks like for @maxfornacon it works fine.
It works for a long time, but eventually the stream gets closed. 

I receive those errors from `snapshot.error` of a StreamBuilder.
```
RealtimeSubscribeException(status: RealtimeSubscribeStatus.channelError, details: RealtimeCloseEvent(code: 1000, reason: heartbeat timeout))
```
and 
```
RealtimeSubscribeException(status: RealtimeSubscribeStatus.timedOut, details: null)
```

>  I think it's better to add them to the .stream() stream, so you can react when the connection is interrupted.

Yes, but there are a lot of the exceptions, but the stream seams to work fine despite of them. Wouldn't handling them cause more harm than good?

Handling the above mentioned exceptions would be fine, because then the stream is really closed and it makes sense to show it to the user to manually reload or implement some kind of auto refresh.

Vinzent03 commented:
The heartbeat exception is thrown from the client side, because the server doesn't answer or similar. But this shouldn't cause the stream to be closed. Adding exceptions to the stream doesn't mean it's closed. The connection of the realtime client may be interrupted, but the stream shouldn't be closed yet. You can determine when a stream is closed by the `onDone` callback in `theStream.listen()`. 
From the package side, there should be no uncaught exceptions. They should all be caught and then added to the stream. You are then supposed to catch them by adding the `onError` and `onDone` callback when using `theStream.listen()`.
https://api.flutter.dev/flutter/dart-async/Stream/listen.html

dshukertjr commented on packages/realtime_client/lib/src/realtime_channel.dart:
```diff
@@ -646,6 +646,19 @@ class RealtimeChannel {
     joinPush.resend(timeout ?? _timeout);
   }
 
+  /// Usually a rejoin only happens when the channel timeouts or errors out.
+  /// When manually disconnecting, the channel is still marked as
+  /// [ChannelStates.joined]. Calling [RealtimeClient.leaveOpenTopic] will
+  /// unsubscribe itself, which causes issues when trying to rejoin. This method
+  /// therefore doesn't call [RealtimeClient.leaveOpenTopic].
+  void forceRejoin([Duration? timeout]) {
```
Could we add `@internal` annotation here? Could we update the comments to describe what this method does as well?

dshukertjr reviewed (commented):
Just had a little comment, but looks awesome!
